util/container/image.go

```diff
@@ -0,0 +1,142 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"fmt"
+	"path"
+	"regexp"
+
+	//  Import the crypto sha256 algorithm for the docker image parser to work
+	_ "crypto/sha256"
+	//  Import the crypto/sha512 algorithm for the docker image parser to work with 384 and 512 sha hashes
+	_ "crypto/sha512"
+
+	"github.com/docker/distribution/reference"
+	"github.com/pkg/errors"
+)
+
+var (
+	ociTagAllowedChars = regexp.MustCompile(`[^-a-zA-Z0-9_\.]`)
+)
+
+// Image type represents the container image details
+type Image struct {
+	Repository string
+	Name       string
+	Tag        string
+	Digest     string
+}
+
+// ImageFromString parses a docker image string into three parts: repo, tag and digest.
+// If both tag and digest are empty, a default image tag will be returned.
+func ImageFromString(image string) (Image, error) {
+	named, err := reference.ParseNamed(image)
+	if err != nil {
+		return Image{}, fmt.Errorf("couldn't parse image name: %v", err)
+	}
+
+	var repo, tag, digest string
+	_, nameOnly := path.Split(reference.Path(named))
+	if nameOnly != "" {
+		// split out the part of the name after the last /
+		lenOfCompleteName := len(named.Name())
+		repo = named.Name()[:lenOfCompleteName-len(nameOnly)-1]
+	}
+
+	tagged, ok := named.(reference.Tagged)
+	if ok {
+		tag = tagged.Tag()
+	}
+
+	digested, ok := named.(reference.Digested)
+	if ok {
+		digest = digested.Digest().String()
+	}
+
+	return Image{Repository: repo, Name: nameOnly, Tag: tag, Digest: digest}, nil
+}
+
+func (i Image) String() string {
+	// repo/name [ ":" tag ] [ "@" digest ]
+	ref := fmt.Sprintf("%s/%s", i.Repository, i.Name)
+	if i.Tag != "" {
+		ref = fmt.Sprintf("%s:%s", ref, i.Tag)
+	}
+	if i.Digest != "" {
+		ref = fmt.Sprintf("%s@%s", ref, i.Digest)
+	}
+	return ref
+}
+
+// ModifyImageRepository takes an imageName (e.g., repository/image:tag), and returns an image name with updated repository
+func ModifyImageRepository(imageName, repositoryName string) (string, error) {
+	image, err := ImageFromString(imageName)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to parse image name")
+	}
+	nameUpdated, err := reference.WithName(path.Join(repositoryName, image.Name))
+	if err != nil {
+		return "", errors.Wrap(err, "failed to update repository name")
+	}
+	if image.Tag != "" {
+		retagged, err := reference.WithTag(nameUpdated, image.Tag)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to parse image tag")
+		}
+		return reference.FamiliarString(retagged), nil
+	}
+	return "", errors.New("image must be tagged")
+}
+
+// ModifyImageTag takes an imageName (e.g., repository/image:tag), and returns an image name with updated tag
+func ModifyImageTag(imageName, tagName string) (string, error) {
```

Comment:
Out of curiosity why are these functions called `Modify*`? Should they just be called `ImageTag` and `ImageRepository` since they return those values respectively?

Reply:
Nevermind I saw the tests below which cleared this up.

Reply:
I think just saying `ImageTag` could imply the func is just returning a value, buts its actually modifying the value from imageName with the new tag

Reply:
Yeah...my bad. I didn't understand at first what the output was but then I saw the tests and it made sense. Thanks! 🙂